File: Diseno de muros concreto/Diseño de muros concreto V2/Guardar_Archivo.vb
Imports System.IO

Public Class Guardar_Archivo

    Dim Ruta As String
    Dim Lista_Texto As List(Of String)

    Public Sub New(ByVal Ruta_archivo As String, ByVal Data_alzado As DataGridView)

        Ruta = Ruta_archivo



        If Muros_lista_2 IsNot Nothing Then
            Actualizar_Resumen()

            Try
                Add_Refuerzoi(data_info_f3)
                Actualizar_Info_ref()
                Actualizar_Alzado()
                Actualizar_long_Alzados()
                Sobre_Escribir()
            Catch ex As Exception
                Actualizar_Info_ref()
                Actualizar_Alzado()
                Actualizar_long_Alzados()
                Sobre_Escribir()
            End Try

        End If

    End Sub

    Sub Actualizar_Resumen()

        Lista_Texto = New List(Of String)
        Dim Texto As String

        For Each muro_i As Muros_Consolidados In Muros_lista_2

            For i = 0 To muro_i.Stories.Count - 1
                Texto = muro_i.Stories(i) & vbTab & muro_i.Pier_name & vbTab & muro_i.lw(i) & vbTab & muro_i.Bw(i) & vbTab & muro_i.fc(i) & vbTab & muro_i.Rho_T(i) & vbTab & muro_i.Rho_l(i) & vbTab & muro_i.Malla(i) & vbTab & muro_i.C_Def(i) & vbTab & muro_i.Lebe_Izq(i) & vbTab & muro_i.Lebe_Der(i) & vbTab & muro_i.Lebe_Centro(i) & vbTab & muro_i.Zc_Izq(i) & vbTab & muro_i.Zc_Der(i) & vbTab & muro_i.Hw(i) & vbTab & muro_i.Est_ebe(i) & vbTab & muro_i.Sep_ebe(i) & vbTab & muro_i.Est_Zc(i) & vbTab & muro_i.Sep_Zc(i) & vbTab & muro_i.As_Long(i) & vbTab & muro_i.ramas_izq(i) & vbTab & muro_i.ramas_der(i) & vbTab & muro_i.ramas_centro(i) & vbTab & muro_i.As_htal(i) & vbTab & muro_i.Ref_htal(i) & vbTab & muro_i.Capas_htal(i) & vbTab & muro_i.sep_htal(i) & vbTab & muro_i.As_Htal_Total(i)
                Lista_Texto.Add(Texto)
            Next

        Next

    End Sub

    Sub Add_Refuerzoi(ByVal Data_info As DataGridView)

        Dim Datos_refuerzo As Refuerzo_muros
        Dim indice As Integer
        Dim suma As Double

        For i = 0 To Data_info.Rows.Count - 1
            Datos_refuerzo = New Refuerzo_muros With {
                .piername = Data_info.Rows(i).Cells(0).Value,
                .pierstory = Data_info.Rows(i).Cells(1).Value,
                .bw = Data_info.Rows(i).Cells(2).Value,
                .as_req = Data_info.Rows(i).Cells(6).Value
            }

            For j = 9 To Data_info.ColumnCount - 1
                Datos_refuerzo.diametro.Add(Int(Mid(Data_info.Columns(j).HeaderText, 2)))
                Datos_refuerzo.cantidad.Add(Data_info.Rows(i).Cells(j).Value)
            Next

            suma = 0
            For j = 0 To Datos_refuerzo.diametro.Count - 1
                suma = suma + areas_refuerzo(Datos_refuerzo.diametro(j)) * Datos_refuerzo.cantidad(j)
            Next

            Datos_refuerzo.total = suma
            Datos_refuerzo.porcentaje = Datos_refuerzo.total / Datos_refuerzo.as_req

            If refuerzo_lista.Count = 0 Or refuerzo_lista.Exists(Function(x) x.piername = Datos_refuerzo.piername And x.pierstory = Datos_refuerzo.pierstory) = False Then
                refuerzo_lista.Add(Datos_refuerzo)
            Else
                indice = refuerzo_lista.FindIndex(Function(x) x.piername = Datos_refuerzo.piername And x.pierstory = Datos_refuerzo.pierstory)
                refuerzo_lista(indice) = Datos_refuerzo
            End If

        Next

    End Sub

    Sub Actualizar_Info_ref()

        Dim Texto As String

        For Each muro_i As Refuerzo_muros In refuerzo_lista

            Texto = muro_i.piername & vbTab & muro_i.pierstory & vbTab & muro_i.bw & vbTab & muro_i.rho & vbTab & muro_i.as_req & vbTab

            For j = 0 To muro_i.diametro.Count - 1
                If j < muro_i.diametro.Count - 1 Then
                    Texto = Texto & muro_i.diametro(j) & vbTab & muro_i.cantidad(j) & vbTab
                Else
                    Texto = Texto & muro_i.diametro(j) & vbTab & muro_i.cantidad(j)
                End If
            Next

            Lista_Texto.Add(Texto)
        Next

    End Sub

    Sub Actualizar_Alzado()
        Dim texto As String
        Dim indice As Integer

        alzado_lista.OrderBy(Function(x) x.pier)
        alzado_lista.OrderBy(Function(x) x.story)
        indice = Lista_Texto.FindIndex(Function(x) x.Contains("Datos de alzado refuerzo longitudinal"))

        If indice < 0 Then
            Lista_Texto.Add("7.Datos de alzado refuerzo longitudinal")
        End If

        For Each Alzado_i As alzado_muro In alzado_lista

            texto = Alzado_i.pier & vbTab & Alzado_i.story & vbTab

            For j = 0 To Alzado_i.alzado.Count - 1
                If j < Alzado_i.alzado.Count - 1 Then
                    texto = texto & Alzado_i.alzado(j) & vbTab
                Else
                    texto = texto & Alzado_i.alzado(j)
                End If
            Next
            Lista_Texto.Add(texto)
        Next
    End Sub

    Sub Actualizar_long_Alzados()

        Dim texto, texto2 As String
        Dim indice, pos As Integer

        alzado_lista.OrderBy(Function(x) x.pier)
        alzado_lista.OrderBy(Function(x) x.story)
        indice = Lista_Texto.FindIndex(Function(x) x.Contains("8.Datos de alzado refuerzo longitudinal - Longitud"))

        If indice < 0 Then
            Lista_Texto.Add("8.Datos de alzado refuerzo longitudinal - Longitud")
        End If

        For Each Alzado_i As alzado_muro In alzado_lista

            texto = Alzado_i.pier & vbTab & Alzado_i.story & vbTab

            For j = 0 To Alzado_i.Alzado_Longitud.Count - 1
                pos = Alzado_i.Alzado_Longitud(j).IndexOf("L=") + 2

                If pos > 2 Then
                    texto2 = Alzado_i.Alzado_Longitud(j).Substring(pos)
                Else
                    texto2 = Alzado_i.Alzado_Longitud(j)
                End If

                If j < Alzado_i.Alzado_Longitud.Count - 1 Then
                    texto = texto & texto2 & vbTab
                Else
                    texto = texto & texto2
                End If
            Next
            Lista_Texto.Add(texto)
        Next

    End Sub

    Sub Sobre_Escribir()

        Dim Inicio, Fin As Integer
        Dim Vector_Texto_aux As New List(Of String)
        Dim sline As String
        Dim X1, indice As Integer
        Dim Lector As New StreamReader(Ruta_1)
        Dim Escritor As StreamWriter

        Do
            sline = Lector.ReadLine()
            Vector_Texto_aux.Add(sline)

        Loop Until sline Is Nothing
        Lector.Close()

        Inicio = Vector_Texto_aux.FindIndex(Function(x) x.Contains("5.Reporte")) + 2

        ''Guardar resumen de diseño
        Try
            Fin = Vector_Texto_aux.FindIndex(Function(x) x.Contains("6.Datos de Refuerzo Adicional")) - 2
        Catch ex As Exception
            Fin = Vector_Texto_aux.FindIndex(Function(x) x.Contains("Fin")) - 2
        End Try

        For i = Inicio To Fin
            Vector_Texto_aux(i) = Lista_Texto(X1)
            X1 += 1
        Next

        ''
        Vector_Texto_aux(Fin + 2) = "6.Datos de Refuerzo Adicional"
        indice = Lista_Texto.FindIndex(Function(x) x.Contains("7.Datos de alzado refuerzo longitudinal"))

        For i = X1 To indice - 1
            Vector_Texto_aux.Add(Lista_Texto(i))
            X1 += 1
        Next

        Vector_Texto_aux.AddRange({"", "7.Datos de alzado refuerzo longitudinal", ""})
        indice = Lista_Texto.FindIndex(Function(x) x.Contains("8.Datos de alzado refuerzo longitudinal - Longitud"))

        For i = X1 + 1 To indice - 1
            Vector_Texto_aux.Add(Lista_Texto(i))
            X1 += 1
        Next


        Vector_Texto_aux.AddRange({"", "8.Datos de alzado refuerzo longitudinal - Longitud", ""})

        For i = indice + 1 To Lista_Texto.Count - 1
            Vector_Texto_aux.Add(Lista_Texto(i))
        Next

        Vector_Texto_aux.AddRange({"", "Fin"})
        Escritor = New StreamWriter(Ruta_1)

        For i = 0 To Vector_Texto_aux.Count - 1
            Escritor.WriteLine(Vector_Texto_aux(i))
        Next


        Escritor.Close()

    End Sub

End Class

File: Analisis muros concreto fc/Diseno muros concreto fc/Guardar archivo.cs
using System.Collections.Generic;
using System.IO;
using System.Windows.Forms;

namespace Diseno_muros_concreto_fc
{
    class Guardar_archivo
    {
        private static List<string> Lista_texto;

        public static void Generar_texto()
        {

            try
            {
                StreamWriter escritor = new StreamWriter(Listas_Programa.Ruta_archivo);

                //Almacenar la capacidad de disipacion de energia
                escritor.WriteLine("0.Capacidad");
                escritor.WriteLine("");
                escritor.WriteLine(Listas_Programa.Capacidad);
                escritor.WriteLine("");

                //Escritor de pier forces
                escritor.WriteLine("1.Wall Forces");
                escritor.WriteLine("");
                Texto_Pier_Forces();

                for (int i = 0; i < Lista_texto.Count; i++)
                {
                    escritor.WriteLine(Lista_texto[i]);
                }

                //Escritor Geometria
                escritor.WriteLine("");
                escritor.WriteLine("2.Wall Geometry");
                escritor.WriteLine("");
                Texto_Geometria();
                for (int i = 0; i < Lista_texto.Count; i++)
                {
                    escritor.WriteLine(Lista_texto[i]);
                }

                //Escritor diseño a cortante
                escritor.WriteLine("");
                escritor.WriteLine("3.Shear Design");
                escritor.WriteLine("");
                Texto_shear_design();

                for (int i = 0; i < Lista_texto.Count; i++)
                {
                    escritor.WriteLine(Lista_texto[i]);
                }

                //Escritor flexural Stress
                escritor.WriteLine("");
                escritor.WriteLine("4.Flexural Stress");
                escritor.WriteLine("");
                Texto_Flexural_Stress();

                for (int i = 0; i < Lista_texto.Count; i++)
                {
                    escritor.WriteLine(Lista_texto[i]);
                }

                //Escritor reporte de diseno
                escritor.WriteLine("");
                escritor.WriteLine("5.Reporte");
                escritor.WriteLine("");
                Texto_reporte();

                for (int i = 0; i < Lista_texto.Count; i++)
                {
                    escritor.WriteLine(Lista_texto[i]);
                }
                escritor.WriteLine("");
                escritor.WriteLine("Fin");
                escritor.Close();
            }
            catch
            {

            }
        }

        public static void Crear_Archivo_Texto()
        {
            string Ruta_archivo;
            SaveFileDialog Guardar_archivo = new SaveFileDialog
            {
                Title = "Guardar archivo de diseño de muros de concreto",
                Filter = "Guardar Archivo |*.dmc"
            };
            Guardar_archivo.ShowDialog();

            Ruta_archivo = Guardar_archivo.FileName;

            Listas_Programa.Ruta_archivo = Ruta_archivo;
            try
            {

                int FinPunto = 0;
                int FinSla = 0;
                string Ruta_Carpeta = "";

                for (int n = 0; n < (Guardar_archivo.FileName.Length); n++) { if (Guardar_archivo.FileName.Substring(n, 1) == ".") { FinPunto = n; } }
                for (int n = FinPunto; n >= 0; n--) { if (Guardar_archivo.FileName.Substring(n, 1) == @"\") { FinSla = n; break; } }
                for (int n = 0; n < FinSla; n++) { Ruta_Carpeta = Ruta_Carpeta + (Guardar_archivo.FileName.Substring(n, 1)); }

                Listas_Programa.Ruta_Carpeta = Ruta_Carpeta;
            }
            catch
            {


            }
        }

        private static void Texto_Pier_Forces()
        {
            Lista_texto = new List<string>();
            string Texto;

            if (Listas_Programa.Lista_Muros is null == false)
            {
                foreach (Muro Muro_i in Listas_Programa.Lista_Muros)
                {
                    for (int j = 0; j < Muro_i.P.Count; j++)
                    {
                        Texto = Muro_i.Story + "\t" + Muro_i.Pier + "\t" + Muro_i.Load[j] + "\t" + Muro_i.Loc[j]
                        + "\t" + Muro_i.P[j] + "\t" + Muro_i.V2[j] + "\t" + Muro_i.V3[j] + "\t" + Muro_i.M2[j] + "\t" + Muro_i.M3[j];
                        Lista_texto.Add(Texto);
                    }
                }

            }


        }

        private static void Texto_Geometria()
        {
            Lista_texto = new List<string>();
            string Texto;

            if (Listas_Programa.Lista_Muros is null == false)
            {
                foreach (Muro Muro_i in Listas_Programa.Lista_Muros)
                {
                    Texto = Muro_i.Story + "\t" + Muro_i.Pier + "\t" + Muro_i.lw + "\t" + Muro_i.bw + "\t" + Muro_i.Fc
                        + "\t" + Muro_i.hw + "\t" + Muro_i.h_acumulado + "\t" + Muro_i.Rho_l_Def;
                    Lista_texto.Add(Texto);
                }
            }
        }

        private static void Texto_shear_design()
        {
            Lista_texto = new List<string>();
            string Texto;

            if (Listas_Programa.Lista_Muros is null == false)
            {
                foreach (Muro Muro_i in Listas_Programa.Lista_Muros)
                {
                    for (int j = 0; j < Muro_i.P.Count; j++)
                    {
                        Texto = Muro_i.Story + "\t" + Muro_i.Pier + "\t" + Muro_i.lw + "\t" + Muro_i.bw + "\t" + Muro_i.Fc + "\t" + Muro_i.h_acumulado + "\t"
                            + Muro_i.Load[j] + "\t" + Muro_i.P[j] + "\t" + Muro_i.V2[j] + "\t" + Muro_i.M3[j] + "\t" + Muro_i.Phi_Vc[j] + "\t"
                            + Muro_i.Phi_Vn_Max1 + "\t" + Muro_i.Phi_Vn_Max2[j] + "\t" + Muro_i.Phi_Vs[j] + "\t" + Muro_i.Phi_Vs_Max + "\t" + Muro_i.Pt_max + "\t"
                            + Muro_i.pt_definitivo[j] + "\t" + Muro_i.Rho_l_Def + "\t" + Muro_i.Cortinas[j] + "\t" + Muro_i.Error_Cortante[j];
                        Lista_texto.Add(Texto);
                    }
                }

            }


        }

        private static void Texto_Flexural_Stress()
        {
            Lista_texto = new List<string>();
            string Texto;

            if (Listas_Programa.Lista_Muros is null == false)
            {
                foreach (Muro Muro_i in Listas_Programa.Lista_Muros)
                {
                    for (int j = 0; j < Muro_i.P.Count; j++)
                    {
                        Texto = Muro_i.Story + "\t" + Muro_i.Pier + "\t" + Muro_i.lw + "\t" + Muro_i.bw + "\t" + Muro_i.Fc + "\t" + Muro_i.h_acumulado + "\t"
                            + Muro_i.Load[j] + "\t" + Muro_i.P[j] + "\t" + Muro_i.M3[j] + "\t" + Muro_i.Fa[j] + "\t" + Muro_i.Fv[j] + "\t" + Muro_i.Sigma_Max[j] + "\t" + Muro_i.Sigma_Min[j]
                            + "\t" + Muro_i.Relacion[j] + "\t" + Muro_i.C_def[j] + "\t" + Muro_i.L_Conf[j] + "\t" + Muro_i.Error_Flexion[j];
                        Lista_texto.Add(Texto);
                    }
                }

            }
        }

        private static void Texto_reporte()
        {
            Lista_texto = new List<string>();
            string Texto;

            if (Listas_Programa.Muros_Consolidados_Listos is null == false)
            {
                foreach (Muros_Consolidados Muro_i in Listas_Programa.Muros_Consolidados_Listos)
                {
                    for (int i = 0; i < Muro_i.Stories.Count; i++)
                    {
                        Texto = Muro_i.Stories[i] + "\t" + Muro_i.Pier_name + "\t" + Muro_i.lw[i] + "\t" + Muro_i.Bw[i] + "\t" + Muro_i.fc[i] + "\t" + Muro_i.Rho_T[i]
                            + "\t" + Muro_i.Rho_l[i] + "\t" + Muro_i.Malla[i] + "\t" + Muro_i.C_Def[i] + "\t" + Muro_i.Lebe_Izq[i] + "\t" + Muro_i.Lebe_Der[i] + "\t" + Muro_i.Lebe_Centro[i]
                            + "\t" + Muro_i.Zc_Izq[i] + "\t" + Muro_i.Zc_Der[i] + "\t" + Muro_i.Hw[i] + "\t" + Muro_i.Est_ebe[i] + "\t" + Muro_i.Sep_ebe[i]
                            + "\t" + Muro_i.Est_Zc[i] + "\t" + Muro_i.Sep_Zc[i] + "\t" + Muro_i.As_Long[i] + "\t" + Muro_i.ramas_izq[i] + "\t" + Muro_i.ramas_der[i]
                            + "\t" + Muro_i.ramas_centro[i] + "\t" + Muro_i.As_htal[i] + "\t" + Muro_i.Ref_htal[i] + "\t" + Muro_i.Capas_htal[i] + "\t" + Muro_i.sep_htal[i]
                            + "\t" + Muro_i.As_Htal_Total[i];
                        Lista_texto.Add(Texto);
                    }
                }

            }

        }
    }
}
